### PR TITLE
Add database/sql driver

### LINKED ISSUE #31: Driver for database/sql
In the other go library it's quite well documented that using the database/sql compatibility wrapper isn't ideal due to how rqlite works. However, for convenience when using existing code and libraries it's very useful to have a wrapper for database/sql while also being aware of the limitations when using it.

For this purpose I wrote a simple wrapper that combines both rqlite and HTTP errors to meaningful error values and tries to keep things fairly minimal in terms of code footprint. I've executed successful database migrations and everything _seems_ to work just fine when writing queries around the limitations of not having transactions in the traditional sense.

The driver does need to do some trickery to avoid issues with blobs and conversions to Go types but with those in place it's honestly quite okay to use.

Would you consider adding a database/sql driver implementation with big warning signs? I'd be happy to cleanup and PR what I have if so.

### PR DESCRIPTION
**Pull request by hifi**
The driver itself is loosely based on gorqlite but is fairly flexible in terms of how it's configured and has quite extensive tests to make sure all basic data types go through correctly in both directions. Errors are categorized into two, request and query errors to be able to know from the caller side if the query failed due to transport issues or a problem in the query itself.

Tests are mostly agent written but every test itself has been requested by me specifically to test some possible edge cases I was expecting to find so they were massaged quite a bit to get to the final end result.

One oddity that might stand out is the casting of `json.Number` to string for scanner. It is required for automatic boolean conversions from integers which is usually expected by users of database drivers. One can also use `json.Number` directly to do their own conversions if they so desire as the string cast works both ways transparently.

I've tested this against [mautrix/go-util/dbutil](https://pkg.go.dev/maunium.net/go/mautrix/util/dbutil) which is the primary library we use internally to access databases in Go but this implementation is completely generic and should work fine as-is as well.

One major motivator to implement this driver was to be able to use the simple and effective database migrations that the dbutil library provides making it possible to version rqlite database schemas with ease in addition to other quality-of-life things we get through it.

ErrorOnBegin option was added as an optional safeguard for unsafe use but it is not enabled by default to allow wide variety of libraries to use this driver without modifications. HTTP basic authentication credential passing through the driver struct is also completely optional if the user of the library wants to have credentials configured separately from the DSN.

No default driver is registered when importing `stdlib` directly so the user of the library can use a standard driver name regardless how they choose to configure it internally. To extend configuration further, a callback based approach like with go-sqlite3 could be used as well so the user gets direct access to the client.

Closes #31

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -22,7 +22,7 @@ jobs:
       RQLITE_GO_HTTP_E2E_HOST: rqlite
     steps:
       - checkout
-      - run: go test -run Test_EndToEnd
+      - run: go test -run Test_EndToEnd ./...
```
Comment (otoolep) [inline .circleci/config.yml:25]: This change is incorrect.

Comment (otoolep) [reply .circleci/config.yml:25]: This line is just about running a specific test, not descending to run all tests too.

Comment (hifi) [reply .circleci/config.yml:25]: The database/sql driver tests are also E2E and have the same prefix.

Do you want a separate test run for the E2E tests that target the stdlib one?

Comment (otoolep) [reply .circleci/config.yml:25]: I see, we can come back to that.

**Comment by otoolep** [comment]
Thanks for your contribution. This is obviously a major PR, and I'm not sure it's practical for me to review all 1,400 lines and be sure the quality meets the bar for rqlite. I'm not saying it doesn't, it's just a large amount of code.

The bar of this is going to be high to get it merged. Unit tests are a start, but it will need something more - some way to independently this is actually functional at a system level. You mentioned system-level testing you do. I would need a step-by-step guide to reproduce that, ideally running in CI.

Creating a separate repo perhaps, and showing me that end-to-end might be best. But I can't commit to reviewing a 1,400 line diff anytime soon.

**Comment by otoolep** [comment]
WDYT @hifi -- this could be useful to folks, but practically we need to take it step by step a bit.

**Comment by hifi** [comment]
> Thanks for your contribution. This is obviously a major PR, and I'm not sure it's practical for me to review all 1,400 lines and be sure the quality meets the bar for rqlite. I'm not saying it doesn't, it's just a large amount of code.

I you take a look, it's only 250 lines of functional code, all the rest is just tests.

> The bar of this is going to be high to get it merged. Unit tests are a start, but it will need something more - some way to independently this is actually functional at a system level. You mentioned system-level testing you do. I would need a step-by-step guide to reproduce that, ideally running in CI.

There are end-to-end tests in the test suite that's part of this PR which is over 80% of it. They test a lot of edge cases for types and such. We're only using rqlite in internal projects for now so I can't share that code directly but I'm quite sure you could plug this into something that expects go-sqlite3 but doesn't do anything with transactions and it should be functional.

> Creating a separate repo perhaps, and showing me that end-to-end might be best. But I can't commit to reviewing a 1,400 line diff anytime soon.

I can also remove the tests to make it easier to review but I don't see much point in that? All the functionality is in sql.go and it's completely self-contained.

**Comment by otoolep** [comment]
Another way to think about this is the following. A compelling case can be made for a large contribution such as this when it's accompanied by statements such as "we're using this in our own production systems", or "we've successfully used to make migrations." Then the community knows it's worth the investment of engaging with it.

That isn't clear to me yet. Perhaps you build this library in a dedicated repo, start getting practical experience with it, which helps us understand that it's truly working well. Then we could move it into the rqlite GitHub org, and you could have full commit rights.

My concern is that LLMs today can generate large amounts of code, that on the surface appear reasonable and pass tests. But that says little about whether that code is actually substantial and useful.

**Comment by hifi** [comment]
LLMs were used mainly for the tests to help validate the driver works as expected. The driver itself is mostly hand written based on gorqlite but with extra consideration on Go types which was helped greatly by having extensive type tests generated.

If you're not willing to review 250 lines of code that's almost functionally identical to the one in gorqlite I don't see much point in continuing.

Thanks, though.

**Comment by otoolep** [comment]
It's not just 250 lines -- the entire PR must be reviewed. It's reviewing tests to make sure they cover the right test cases, that the test cases are meaningful, well-constructed, and maintainable.

**Comment by otoolep** [comment]
@hifi -- to be clear, I appreciate the contribution. I'm just highlighting that any 1400 line PR is a big PR to review and that presents very practical problems. Helping understand how much this is being used in production, its history, whether it has shown itself to be actual useful helps a lot.

**Comment by hifi** [comment]
While I wholeheartedly agree that carefully reviewing 1400 lines of code is not something I'd like to do either your stance on the tests gives the impression it would have been better not to have them at all because they are too extensive to review by hand which devalues the feature itself from being reviewed or merged.

These tests are what I used myself so they don't necessarily need to be the ones in this repo but for completeness I kept them as is and ported in. If you want me to simplify them to what the main E2E tests are that'd be fine for me but it wouldn't actually test much.

The driver itself is quite small and easy to follow and at best requires little to no maintenance because it's only a thin wrapper.

I can of course keep this out-of-tree but it wouldn't have much exposure that way hence why I thought upstreaming would be the right move if it's in scope for the project.

**Comment by otoolep** [comment]
Thanks @hifi -- let me find some time to take a look at all this, might just be a little while.

**Comment by hifi** [comment]
I spent way too much time on this but to help prove this driver is actually worth it I kept [poking](https://github.com/beeper/kine/commit/9fbe7be660a028239972ce3acfd7e00308abc897) [Kine](https://github.com/k3s-io/kine) until [all](https://github.com/beeper/kine/actions/runs/25957873976/job/76308044963/) etcd/api-server tests passed and the subset of Kubernetes e2e tests as well.

Given the all these tests pass like they would for SQLite I dare to say there can't be anything majorly wrong with this implementation. As a bonus it should be possible to run a Kubernetes cluster on rqlite with minimal changes to Kine.

The CI is slow and flaky so if that specific test run fails before it completes here's the K8s E2E results from my machine:
```
[rqlite] Ran 167 of 7355 Specs in 1578.364 seconds
[rqlite] SUCCESS! -- 167 Passed | 0 Failed | 0 Pending | 7188 Skipped
[rqlite] PASS
```

The Kine changes are not intended to be upstreamed, it was just to get the tests working. It likely makes more sense to have a rqlite specific driver in Kine rather than shoehorning it into the SQLite one.

To fix up Kine I implemented write-only transactions for the driver so it expanded slightly but not too much.
